### PR TITLE
DML EP Register Opset 21

### PR DESCRIPTION
**Pull request by dtang317**
### Description
This PR registers the following opset 21 operators:
- Size-21
- CastLike-21
- ConstantOfShape-21
- Flatten-21
- Pad-21
- Transpose-21

### CODE REVIEW

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp [hunk 1]
```diff
@@ -792,6 +799,7 @@ constexpr static OperatorRegistrationInformation operatorRegistrationInformation
     {REG_INFO_VER( 18,  Split,                              typeNameListDefault,            supportedTypeListAllScalars,            DmlGraphSupport::Supported,      requiredConstantCpuInputs(1))},
     {REG_INFO(      7,  Transpose,                          typeNameListDefault,            supportedTypeListAllScalars,            DmlGraphSupport::Supported)},
     {REG_INFO(     13,  Transpose,                          typeNameListDefault,            supportedTypeListAllScalars,            DmlGraphSupport::Supported)},
+    {REG_INFO(     21,  Transpose,                          typeNameListDefault,            supportedTypeListTranspose21,            DmlGraphSupport::Supported)},
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp:802]:
```suggestion
    {REG_INFO(     21,  Transpose,                          typeNameListDefault,            supportedTypeListTranspose21,           DmlGraphSupport::Supported)},
```

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp [hunk 2]
```diff
@@ -599,6 +601,7 @@ constexpr static std::array<SupportedTensorDataTypes, 2> supportedTypeListMaxPoo
 constexpr static std::array<SupportedTensorDataTypes, 2> supportedTypeListMaxUnpool = {SupportedTensorDataTypes::Float16to32, SupportedTensorDataTypes::Int64};
 constexpr static std::array<SupportedTensorDataTypes, 1> supportedTypeListIndices = { SupportedTensorDataTypes::Int32|SupportedTensorDataTypes::Int64 };
 constexpr static std::array<SupportedTensorDataTypes, 2> supportedTypeListCast = { SupportedTensorDataTypes::AllScalars, SupportedTensorDataTypes::AllScalars };
+constexpr static std::array<SupportedTensorDataTypes, 2> supportedTypeListCast21 = { SupportedTensorDataTypes::AllScalars, SupportedTensorDataTypes::UInt4 | SupportedTensorDataTypes::Int4 };
```
[the remark this answers is not in the document]
Comment (dtang317) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp:604]: I ran local tests except for Cast and Castlike (I didn't see test for castlike?) Cast 21 was registered by someone else (line 1098), yea the supported data type for Cast doesn't contain int4/uint4

[the remark this answers is not in the document]
Comment (fdwr) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp:604]: I have little confidence that the existing test cases (onnxruntime_test_all.exe) cover this specific combo of cast with int4.

[the remark this answers is not in the document]
Comment (fdwr) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp:604]: Actually I do see a combo here: https://github.com/onnx/onnx/blob/main/docs/Operators.md#cast

    ("FLOAT", "INT4"),
    ("FLOAT16", "INT4"),

And https://github.com/onnx/onnx/tree/main/onnx/backend/test/data/node/test_cast_FLOAT_to_UINT4.

So it's a mystery if the ONNX backend test cases pass, even though DML doesn't support int4 as an output. It makes me wonder if something else is up, and if you're actually reaching the DML EP.

Is a breakpoint hit in ORT here:

`onnxruntime\onnxruntime\core\providers\dml\DmlExecutionProvider\src\Operators\DmlOperatorCast.cpp`

or in DML?

`directml\SharedValidation\DmlCastOperatorValidator.h`

**Comment by dtang317** [comment]
@microsoft-github-policy-service agree company="Microsoft"
________________________________

**Review by fdwr — DISMISSED** [review]
👍

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (fdwr) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp]: Deferring this aspect to later, given you removed the int4 data types in the type list.

**Review by fdwr — APPROVED** [review]

**Comment by fdwr** [comment]
@dtang317 If you want review from anybody else, let me know. Otherwise I will merge it (all tests passed).

**Pull request merged**